streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java

```diff
@@ -0,0 +1,691 @@
+
+            // capture committed state for later
+            final ReadOnlyKeyValueStore<Long, Long> storeOneBeforeUpgrade =
+                streams1Alpha.store(StoreQueryParameters.fromNameAndType(storeName, QueryableStoreTypes.keyValueStore()));
+            final Map<Long, Long> currentStateOne = new HashMap<>();
+            final long keyFilterFirstUpgrade;
+            try (final KeyValueIterator<Long, Long> it = storeOneBeforeUpgrade.all()) {
+                keyFilterFirstUpgrade = it.peekNextKey();
+                while (it.hasNext()) {
+                    final KeyValue<Long, Long> row = it.next();
+                    currentStateOne.put(row.key, row.value);
+                }
+            }
+            final ReadOnlyKeyValueStore<Long, Long> storeTwoBeforeUpgrade =
+                streams2Alpha.store(StoreQueryParameters.fromNameAndType(storeName, QueryableStoreTypes.keyValueStore()));
+            final Map<Long, Long> currentStateTwo = new HashMap<>();
+            try (final KeyValueIterator<Long, Long> it = storeTwoBeforeUpgrade.all()) {
+                while (it.hasNext()) {
+                    final KeyValue<Long, Long> row = it.next();
+                    currentStateTwo.put(row.key, row.value);
+                }
+            }
+
+            // phase 3:
+            // expected end state per output partition:
+            // p-0: 10 records + COMMIT + 5 records (pending)
+            // p-1: 10 records + COMMIT + 5 records (pending)
+            stateTransitions2.clear();
+            final AtomicBoolean errorInjectedClient2 = new AtomicBoolean(false);
+            errorInjected = errorInjectedClient2;
+            if (injectError) {
+                errorInjectedClient1.set(true);
+            }
+            final List<KeyValue<Long, Long>> uncommittedInputDataBeforeUpgrade = prepareData(10L, 15L, 0L, 1L);
+            allData.addAll(uncommittedInputDataBeforeUpgrade);
+
+            writeInputData(uncommittedInputDataBeforeUpgrade);
+
+            final List<KeyValue<Long, Long>> allOutputDataBeforeUpgrade = readResult(allData.size(), null);
+            checkResultPerKey(allOutputDataBeforeUpgrade, computeExpectedResult(allData, new HashMap<>()));
+
+            // phase 4:
+            // expected end state per output partition:
+            // stop case:
+            //   p-0: 10 records + COMMIT + 5 records + COMMIT
+            //   p-1: 10 records + COMMIT + 5 records (pending)
+            // crash case:
+            //   p-0: 10 records + COMMIT + 5 records (pending)
+            //   p-1: 10 records + COMMIT + 5 records (pending)
+            if (!injectError) {
+                streams1Alpha.close();
+            }
+            waitForCondition(
+                () -> stateTransitions2.equals(Arrays.asList(
+                    KeyValue.pair(KafkaStreams.State.RUNNING, KafkaStreams.State.REBALANCING),
+                    KeyValue.pair(KafkaStreams.State.REBALANCING, KafkaStreams.State.RUNNING)
```

Comment:
After it resumes to running, p-0 should be aborted right?

Reply:
Well, the original TX for p-0 would be aborted (with zero pending writes; compare my comment above), but client-2 will start a new TX when retrying to process records 10 to 14 of p-0.

So technically it would we:
```
p-0: 10 records + COMMIT + ABORT + 5 records (pending)
```

Shall I update the comment?